### ISSUE TITLE
Segfaults on GPU with particular MasterElement methods

### ISSUE THREAD

**Issue by sayerhs** [Exawind/nalu-wind#186]
Certain [calls](https://github.com/Exawind/nalu-wind/blob/master/include/ScratchViews.h#L622) to MasterElement methods to populate ScratchViews from within device loop causes segfaults. 

Determine and fix cause of segfaults, update `UnitTestHexMasterElementNgp.C` to add unit tests that demonstrate execution of these methods on GPUs.

**Comment by overfelt** [comment]
@alanw0,

I added a branch that has a unit test to show this.  I should not have commented out UnitTestHexMasterElementsNgp.C:193 since this is a branch, but if you clone and uncomment that line and run the unit test with --gtest_filter=masterElementHexSerialNGP.hex8_scs_interpolation you should get an error when trying to set up the element data requests.  I do not see that it has anything to do with the master elements.  It seems to die just allocating a simple Kokkos::View.  

Could you just try to duplicate the test in your project?  

jro

**Comment by alanw0** [comment]
@overfelt I will try to reproduce this.

**Comment by sayerhs** [comment]
@overfelt I had the same experience. The code will segfault in the fill master element views even if you register no ME views in the ElemDataRequests object, which is really puzzling.

**Comment by overfelt** [comment]
I'm trying to build -G to debug through cuda calls.  I could not get Trilinos to build, some odd undefined references but maybe just nalu is enough.

**Comment by alanw0** [comment]
@overfelt I finally tried to reproduce this. I built and ran with line 193 commented out and uncommented, and in both cases it runs without complaining... How can I make sure I'm running the right thing here?

**Comment by alanw0** [comment]
@overfelt I am now reproducing the crash. I had to also uncomment the SharedMemData construction inside the kokkos loop. I'll see what I can figure out about the crash.

**Comment by alanw0** [comment]
The scratchViews change that triggers this error is checked into branch issue_186 of my fork: https://github.com/alanw0/nalu-wind/tree/issue_186

**Comment by overfelt** [comment]
So for the issue: Segfaults on GPU with particular MasterElement methods #186.

This seems to be a problem with deleting Views on the GPU.   During destruction of a View, there is an info structure filled with a call to the cuda library.  On host it returns lots of information but on the GPU it is all zeros.  This is what leads to the divide by zero when trying to delete the view.  Is it possible that views should not be deleted on the GPU?  Probably.  

I do not know why this sometimes works or why adding code in the master elements that is never actually called makes a difference.  It seems that it should always fail. 

One fix is to put an empty destructor in the FieldInfoNGP class so the view it owns is not destroyed.  But this might be a memory leak instead and probably not what we want to do in all cases.  Maybe only destroy the view when running on host?

**Comment by sayerhs** [comment]
@overfelt Can you explain why you think we need to fiddle with `FieldInfoNGP`? the `MasterElement` data never gets stored in `FieldInfoNGP`. Or were you talking about `MultiDimViews`?

**Comment by sayerhs** [comment]
@overfelt James had to revert #379 in d4706afdc5c01aaa because it was causing still failing tests on CDash. It seems that we are still unable to run with the `#ifdefs` removed for those master element data. Can you share what CUDA version + GPU you were using where you saw successful run?

**Comment by overfelt** [comment]
Bummer.  I am running on the ascic platforms here at Sandia. Tesla P100s with nvidia-9.2.88.

**Comment by overfelt** [comment]
I'll do a clean build to make sure.

**Comment by overfelt** [comment]
Well, duh, of course.  I moved the fix into that Simulation class which is not used by unit_test.C.

**Comment by sayerhs** [comment]
Of course! We should have noticed that. Can you please resubmit a new pull request?

**Comment by sayerhs** [comment]
@overfelt Commit d4706afdc5c reverted your fix in `Simulation.C` for setting stack size (9c079cde78). Can you please put that back in so that we can proceed with `naluX` running on GPUs after #422

**Comment by overfelt** [comment]
Your asking to revert the revert?  There must have been a good reason @jrood-nrel reverted that commit but after a month I don't recall what it was.  Does the commit in #422 break the GPU runs again?  The tests seem to run fine for me.

**Comment by sayerhs** [comment]
@overfelt so @jrood-nrel reverted the commit because we were seeing segfaults in unittest, but it turned out that you had forgotten to put the cuda stack size fix in `unit_tests.C`. So I was wondering if we need the stack size fix in `nalu.C` (or `Simulation.C`) also so that we don't get segfaults when running full `naluX` on GPUs.

**Comment by overfelt** [comment]
Ahhh, I see.  Yes, at least it might be easier to diagnose running out of heap space rather than the odd segfaults we get when the stack runs out.  I'll revert the commit.

**Comment by sayerhs** [comment]
Fixed in #427
